src/stages/kinesis-s3.ts

```diff
@@ -0,0 +1,82 @@
+import * as firehose from '@aws-cdk/aws-kinesisfirehose-alpha';
+import * as destinations from '@aws-cdk/aws-kinesisfirehose-destinations-alpha';
+import * as events from 'aws-cdk-lib/aws-events';
+import * as kinesis from 'aws-cdk-lib/aws-kinesis';
+import * as s3 from 'aws-cdk-lib/aws-s3';
+import { Construct } from 'constructs';
+import { DefaultDestinationsS3BucketProps } from '../core/kinesis-firehose-defaults';
+import { DefaultS3BucketProps } from '../core/s3-defaults';
+import { consolidateProps, overrideProps } from '../core/utils';
+import { DataStage, DataStageProps } from '../pipelines/stage';
+
+export interface FirehoseToS3StageProps extends DataStageProps {
+  readonly s3Bucket?: s3.IBucket;
+  readonly s3BucketProps?: s3.BucketProps;
```

Comment:
I have used the existing `BucketProps` and `destinations.S3BucketProps` interfaces instead of defining my own. Is there are reason why we have adopted a different strategy in other stages (i.e. defining our own interfaces for these)?

Reply:
Mostly I was copying these definitions as they were in Python. I figured the reasoning was that we didn't want them messing with some important defaults (like making a function fully open to the public or something). Or if they really did, they would have to pass their own Function/Bucket.

Another concern I have is that when a user reads the documentation for what goes into `FirehoseToS3StageProps`, they'll look at the properties of `s3.BucketProps` and then they'll read that if they don't set the encryption, it will be set to unencrypted. This will be confusing if we're offering different default values, such as `S3_MANAGED` for encryption. A benefit of defining our own set of properties is that at least our documentation will be accurate.

Reply:
Fair points. I suppose my concern about restricting the arguments of Props is that there will always be a user who needs a specific argument that we wouldn't be supporting (c.f. Tracing in the Lambda Function). I find the idea of asking them to create their own Lambda function to be a bad UX and kind of defeats the point of DDK a bit. I suppose it goes back to the discussion on props and escape hatches again...

Reply:
Typescript does let us hack around this a little bit. We could have something like:
```typescript
interface OurBucketProps extends Omit<s3.BucketProps, "encryption"> {
  /**
    Our custom documentation, and here we can mention what our default is.
  */
  readonly encryption: s3.EncryptionOption;
}
```
This way, `OurBucketProps` extends a version of the `BucketProps` that omits the `encryption` property. This allows us to redefine it ourself, with our own documentation, while still inheriting all other properties. And if we then decide to do the same thing with `enforceSsl`, we simply inherit from `Omit<s3.BucketProps, 'encryption', 'enforceSsl'>` , and define our own `enforceSsl`.

This solution would technically solve the problem for us, although my primary concern here is I'm not sure how well this would play with other languages as it gets translated with JSII.

Reply:
A pattern that I've seen is to extend the type even if there are no additions to the type at the moment i.e.
```
interface OurBucketProps extends s3.BucketProps {}
```
This way we can still keep the code backward-compatible:
1. Add properties when required
2. Break the inheritance if we have to divert from the original definition of a property at some point without changing the type of what the user has to provide

Reply:
`Omit` & other utility types are cool but I would avoid them unless absolutely necessary as they're a bit of a hack

Reply:
The problem is that without `Omit`, JSII will not let you redefine a property, even to just rewrite the property documentation. When I tried, it said it would break compatibility with C# (and Java I believe).